### ISSUE TITLE
Drop 2to3 mechanism

### ISSUE THREAD

**Issue by shimizukawa** [sphinx-doc/sphinx#1350]
There is some way to complete to drop 2to3.
- Use [six](https://pypi.python.org/pypi/six)
- Use [python-future](https://pypi.python.org/pypi/future)
- Pure python code.

(see also #1220 discussion)

---
- Bitbucket: https://bitbucket.org/birkenfeld/sphinx/issue/1350
- Originally reported by: Takayuki Shimizukawa
- Originally created at: 2014-01-18T12:03:28.321

**Comment by shimizukawa** [comment]
_From Dmitry Shachnev on 2014-01-18 11:05:09+00:00_

Also:

```
from __future__ import print_function
```

Also: replace `except A, b` with `except A as b`.

**Comment by shimizukawa** [comment]
_From Takayuki Shimizukawa on 2014-01-18 11:10:57+00:00_

[Dmitry Shachnev](https://bitbucket.org/mitya57) Exactly!

**Comment by shimizukawa** [comment]
_From Takayuki Shimizukawa on 2014-01-18 11:28:25+00:00_

The python-future does not support Python-3.2 because py32 does not support unicode literal. They said `Adding support for Python 3.2 to future would likely impose a penalty with performance and/or maintainability,..` (see also [python-future FAQ](http://python-future.org/faq.html#platform-and-version-support)).

I think we need to decide whether drop Python-3.2 or not. If we decided to drop py32, we may not need to be dependent on six/future.

**Comment by shimizukawa** [comment]
_From Georg Brandl on 2014-01-18 11:38:38+00:00_

For that it would be nice to know what `python-future` offers on top of `six`, and if it's worth using it. If yes, I'm not opposed to requiring Python 3.3.

**Comment by shimizukawa** [comment]
_From Georg Brandl on 2014-01-18 11:40:36+00:00_

Independent changes like those Dmitry suggested can of course be made straight away.

**Comment by shimizukawa** [comment]
_From Georg Brandl on 2014-01-18 12:37:40+00:00_

Looking at future, these can be made automatically using "futurize --stage1".

**Comment by shimizukawa** [comment]
_From Georg Brandl on 2014-01-18 15:48:28+00:00_

I've just finished porting Pygments to single-source, and I have two observations:
- python-future's conversion tool is nice, but otherwise I prefer six: less magic. future even reimplements the "int" type which is definitely too much compatibility.
- The u"" literal from Python 3.3+ is not definitely needed; modules with many literals can use `from __future__ import unicode_literals` while others can use a `u()` wrapper.  For Pygments, I did not use `unicode_literals` since it pretty much changes all lexers. Therefore I require 3.3+ there.

**Comment by shimizukawa** [comment]
_From Dmitry Shachnev on 2014-01-19 10:27:56+00:00_

I have just created https://bitbucket.org/birkenfeld/sphinx/pull-request/208 for the first part of refactoring. Please review it carefully, I may have missed some bugs.

I have a question about `tests/etree13`. Is it still needed? Can we use `xml.etree.ElementTree` instead (which is available in Python 2.5+)?

**Comment by shimizukawa** [comment]
_From Georg Brandl on 2014-01-19 10:31:20+00:00_

If the tests work with it, yes. Probably they will.

**Comment by shimizukawa** [comment]
_From Dmitry Shachnev on 2014-01-19 10:32:32+00:00_

Dropping that will be my next pull request, then.

**Comment by shimizukawa** [comment]
_From Dmitry Shachnev on 2014-01-29 14:59:09+00:00_

Re `etree13`, the tests work with Python 2 but not Python 3, I get lots of failures like this one:

```
======================================================================
FAIL: test_build_html.test_html(<xml.etree.ElementTree.ElementTree object at 0xb5f2f3ec>, 'subdir/includes.html', ".//a[[False](https://bitbucket.org/href)='../_downloads/img.png']", '')
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/home/dmitry/sphinx-hg/sphinx/build/lib/tests/test_build_html.py", line 302, in check_xpath
    '%r in file %s' % (path, fname))
AssertionError: did not find any node matching xpath ".//a[[False](https://bitbucket.org/href)='../_downloads/img.png']" in file subdir/includes.html
```

**Comment by shimizukawa** [comment]
_From Takayuki Shimizukawa on 2014-04-26 09:18:48+00:00_

I got a reason why py3 test has been failed.

Python 3 imports compiled version ElementTree and XMLParser automatically.
On the other hand, in Python 2, ElementTree and cElementTree is divided explicitly.

And compiled XMLParser doesn't have `_fixname` method then subclass can't override it. http://hg.python.org/cpython/file/9fc4a1ebe652/Modules/_elementtree.c#l3608

I'm going to introduce tests/pycompat.py that has a full copy of the python version XMLParser. I think it will resolve this problem.

**Comment by shimizukawa** [comment]
_From Takayuki Shimizukawa on 2014-04-26 09:38:57+00:00_

I created a https://bitbucket.org/birkenfeld/sphinx/pull-request/237 to remove tests/etree13. Please review it.

**Comment by shimizukawa** [comment]
_From Takayuki Shimizukawa on 2014-04-30 04:11:28+00:00_

I created a https://bitbucket.org/birkenfeld/sphinx/pull-request/238 that introduce the six module. Please review it.

(FYI, I declined the https://bitbucket.org/birkenfeld/sphinx/pull-request/237 for py26 and py3.)

**Comment by shimizukawa** [comment]
_From Takayuki Shimizukawa on 2014-05-24 16:29:57+00:00_

Merged in shimizukawa/sphinx-use-six (https://bitbucket.org/birkenfeld/sphinx/pull-request/238)

using six package for py2/3 compatibility without 2to3. refs #1350

→ <<cset 80fd0f72d958c57070b35dcd9da33ece6a3589ed>>

**Comment by shimizukawa** [comment]
_From Takayuki Shimizukawa on 2014-05-24 16:29:57+00:00_

use six privided functions/classes to support py2/py3 in one source. refs #1350

→ <<cset f31d9113ffc1c58bfcc47f2a29439102b62801fc>>

**Comment by shimizukawa** [comment]
_From Takayuki Shimizukawa on 2014-05-24 16:29:57+00:00_

use UnicodeMixin for **str**, **unicode** to support py2/py3 in one source. refs #1350.

→ <<cset 9dbb6bf0927b4b04cc95b0ec27ece2325c5c6053>>

**Comment by shimizukawa** [comment]
_From Takayuki Shimizukawa on 2014-05-24 16:29:57+00:00_

use six privided text_type() to replace with unicode() to support py2/py3 in one source. refs #1350.

→ <<cset 00eff0b7f656b5212b6f4d2c567f6828ee81ee82>>

**Comment by shimizukawa** [comment]
_From Takayuki Shimizukawa on 2014-05-24 16:29:57+00:00_

provide **next**() and use native next() to support py2/py3 in one source. refs #1350.

→ <<cset 91b9d75ac88afc349de4f8629b99c7fbcb3ff841>>

**Comment by shimizukawa** [comment]
_From Takayuki Shimizukawa on 2014-05-24 16:29:57+00:00_

wrap iterator with list() to avoid 'dictionary changed size during iteration' on py3. refs #1350.

→ <<cset 49944cfa5229356f7b40a9ec232f11ca5c5393fa>>

**Comment by shimizukawa** [comment]
_From Takayuki Shimizukawa on 2014-05-24 16:29:57+00:00_

use six privided iteritems(),itervalues() to support py2/py3 in one source. refs #1350.

→ <<cset ce4d342e20739ab994a451f367d5bce2c5edb5d2>>

**Comment by shimizukawa** [comment]
_From Takayuki Shimizukawa on 2014-05-24 16:29:57+00:00_

use six privided functions/classes to support py2/py3 in one source. refs #1350.

→ <<cset c3c879f2c63afbb39ad4d76ecc083cddd41d9946>>

**Comment by shimizukawa** [comment]
_From Takayuki Shimizukawa on 2014-05-24 16:29:57+00:00_

introduce the six module and reduce sphinx.util.pycompat implementation. refs #1350.

→ <<cset 68290a266cea33cf0b60bed425b5e89d1032e7d1>>

**Comment by shimizukawa** [comment]
_From Takayuki Shimizukawa on 2014-05-24 16:29:57+00:00_

Merged in shimizukawa/sphinx-use-six (https://bitbucket.org/birkenfeld/sphinx/pull-request/238)

using six package for py2/3 compatibility without 2to3. refs #1350

→ <<cset 80fd0f72d958c57070b35dcd9da33ece6a3589ed>>

**Comment by shimizukawa** [comment]
_From Takayuki Shimizukawa on 2014-05-24 16:37:28+00:00_

I created a https://bitbucket.org/birkenfeld/sphinx/pull-request/243 as native support py2/py3 without 2to3. Please review it.

**Comment by shimizukawa** [comment]
_From Takayuki Shimizukawa on 2014-05-27 14:01:42+00:00_

Merged in shimizukawa/sphinx-py3-native (https://bitbucket.org/birkenfeld/sphinx/pull-request/243)

native py2/py3 support without 2to3. refs #1350

→ <<cset a335414b8176b269181fe0d5fd2d9906b1a94938>>

**Comment by shimizukawa** [comment]
_From Takayuki Shimizukawa on 2014-05-27 14:01:42+00:00_

fix py2/py3 compat for 'iter.next()/next(iter)' and 'dict.values()/list(dict.values())'. refs #1350.

→ <<cset 5fd8dc885f4851b6a84be8817dff99508f328e25>>

**Comment by shimizukawa** [comment]
_From Takayuki Shimizukawa on 2014-05-27 14:01:42+00:00_

use 'next(iter)' instead of 'iter.next()' to support py2/py3 compatibiity. refs #1350.

→ <<cset 071fb45a5434c7d0928fddacd486363e05ce920c>>

**Comment by shimizukawa** [comment]
_From Takayuki Shimizukawa on 2014-05-27 14:01:42+00:00_

wrap py3 iterators with list() for each places that expect a list object. refs #1350.

→ <<cset 2d1549b35a4adea66aebd0b7d9266731dd7ece6f>>

**Comment by shimizukawa** [comment]
_From Takayuki Shimizukawa on 2014-05-27 14:01:42+00:00_

replace ur'' prefix with r'' to support py2/py3 in one source. refs #1350

→ <<cset 6ae3b68859562f07a1a7e215a7b55dce800305a9>>

**Comment by shimizukawa** [comment]
_From Takayuki Shimizukawa on 2014-05-27 14:01:42+00:00_

Merged in shimizukawa/sphinx-py3-native (https://bitbucket.org/birkenfeld/sphinx/pull-request/243)

native py2/py3 support without 2to3. refs #1350

→ <<cset a335414b8176b269181fe0d5fd2d9906b1a94938>>

**Comment by shimizukawa** [comment]
_From Takayuki Shimizukawa on 2014-05-27 14:23:56+00:00_
- Drop Python-3.2. And now, removing 2to3 mechanism procedure has been completed. Closes #1350.

Pygments-2.0 will drop Python-3.2.
The snowballstemmer package that is referred from sphinx-1.3 (that is merged at https://bitbucket.org/birkenfeld/sphinx/pull-request/214) has not supported Python-3.2.

We decided to drop the Python-3.2.

see also https://bitbucket.org/birkenfeld/sphinx/pull-request/243/native-py2-py3-support-without-2to3-refs

→ <<cset 6746c83a188e208041e0463c4e3fc5fb15316b07>>
